frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseResetBranchToRemoteAction.java

```diff
@@ -0,0 +1,188 @@
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendActions");
+  private static final String VCS_LOGGER_TITLE = "Resetting";
+  private static final String TASK_TITLE = "Resetting...";
+
+  protected abstract Option<String> getBranchName(AnActionEvent anActionEvent);
+
+  protected abstract String getActionType();
+
+  @Override
+  @UIEffect
+  public void update(AnActionEvent anActionEvent) {
+    super.update(anActionEvent);
+
+    Presentation presentation = anActionEvent.getPresentation();
+    if (!presentation.isEnabledAndVisible()) {
+      return;
+    }
+
+    Option<String> branchNameOption = getBranchName(anActionEvent);
+
+    if (branchNameOption.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled due to undefined ${getActionType()} branch");
+      return;
+    }
+
+    String branchName = branchNameOption.get();
+
+    Option<SyncToRemoteStatus> syncToRemoteStatus = getGitMacheteRepository(anActionEvent)
+        .flatMap(repo -> repo.getBranchByName(branchName))
+        .map(BaseGitMacheteBranch::getSyncToRemoteStatus);
+
+    if (syncToRemoteStatus.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled due to undefined sync to remote status");
+      return;
+    }
+
+    SyncToRemoteStatus.Relation relation = syncToRemoteStatus.get().getRelation();
+
+    if (relation != SyncToRemoteStatus.Relation.Untracked) {
+      presentation.setDescription("Reset branch '${branchName} to its remote tracking branch");
+    } else {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled because ${getActionType()} branch is untracked");
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    LOG.debug("Performing");
+
+    Project project = getProject(anActionEvent);
+    Option<GitRepository> selectedVcsRepository = getSelectedVcsRepository(anActionEvent);
+    Option<String> branchName = getBranchName(anActionEvent);
+    Option<IGitMacheteRepository> macheteRepository = getGitMacheteRepository(anActionEvent);
+
+    if (branchName.isDefined()) {
+      if (selectedVcsRepository.isDefined()) {
+        if (macheteRepository.isDefined()) {
+          doResetToRemoteWithKeep(project, selectedVcsRepository.get(), branchName.get(), macheteRepository.get(),
+              anActionEvent);
+        } else {
+          LOG.error("Skipping the action because can't get Git Machete repository");
+          VcsNotifier.getInstance(project).notifyError(VCS_LOGGER_TITLE,
+              "Internal error occurred. For more information see IDE log file");
+        }
+      } else {
+        LOG.warn("Skipping the action because no VCS repository is selected");
+        VcsNotifier.getInstance(project).notifyWarning(VCS_LOGGER_TITLE,
+            "Skipping the action because no VCS repository is selected");
```

Comment:
This is an inconsistency to the messages above and below. 
And I am not convinced that `Internal error occurred. For more information see IDE log file` is something that we wanna show. I think that an information with a cause like `because name of branch to reset is undefined` would be more helpfull to the user. 

Feel free to open an issue to clean up and provide some conventions to vcs notification and logs within the scope of `frontendActions`.

Reply:
Created #320